### ISSUE TITLE
Invalid diagnostic location for a sub-call to a specialized `ParamSpec`

### ISSUE THREAD

**Issue by dhruvmanila** [astral-sh/ty#2198]
### Summary

Given the following example:
```py
from typing import Callable

def fn(a: int) -> None: ...

def foo[**P, T](fn: Callable[P, T], *args: P.args, **kwargs: P.kwargs): ...

foo(fn, a="a")
```

The diagnostic is using the location for `fn` for highlighting on `foo`:

```console
❯ ty check .                                                          
error[invalid-argument-type]: Argument to function `foo` is incorrect
  --> main.py:10:5
   |
10 | foo(fn, a="a")
   |     ^^ Expected `int`, found `Literal["a"]`
   |
info: Function defined here
 --> main.py:7:5
  |
7 | def foo[**P, T](fn: Callable[P, T], *args: P.args, **kwargs: P.kwargs): ...
  |     ^^^         ------------------ Parameter declared here
  |
info: rule `invalid-argument-type` is enabled by default

Found 1 diagnostic
```

1. The "Expected `...`, found `...`" should be on `"a"`
2. It should use the `fn` function directly for the info diagnostic of "Function defined here"
3. And, provide a sub-diagnostic / info to state that this function comes from the argument being matched against a `ParamSpec` type variable which resolved to this function

**Comment by 11happy** [comment]
Hello @dhruvmanila , I would like to work on this.
Thank you : )

**Comment by dhruvmanila** [comment]
Go for it! I think (1) should be relatively easy to implement, we need to offset the error location with the number of parameters while evaluating the paramspec sub-call in `evaluate_paramspec_sub_call`. For (2) and (3), maybe we can start with avoiding the "Function defined here" sub-diagnostic and just add the `Callable` type as the diagnostic information (`info: ...`).

**Comment by 11happy** [comment]
I have created the PR fixing first part , to keep it clean as second & third part would probably require to change structs to track original function location, I would create a follow up PR for that.
Thank you : )
